File: 0957-minimum-add-to-make-parentheses-valid/0957-minimum-add-to-make-parentheses-valid.py
class Solution:
    def minAddToMakeValid(self, s: str) -> int:
        # The Greedy One
        # '(' opening++
        #  ')' opening-- and closing_imbalance++

        # O(n) and O(1)
        opening = 0
        closing_imbalance = 0

        for c in s:
            if c == '(':
                opening += 1
            else:
                if opening == 0:
                    closing_imbalance += 1
                else:
                    opening -= 1
        
        return opening + closing_imbalance
        

File: 0957-minimum-add-to-make-parentheses-valid/0957-minimum-add-to-make-parentheses-valid.java
class Solution {
    public int minAddToMakeValid(String s) {
        if(s.length() == 1) return 1;
        if(s.length() == 0) return 0;
        Stack<Character> open = new Stack<>();
        // wrong answer- "))(("
        int min_moves = s.length();
        for(char c : s.toCharArray()) {
            if(c == '(') {
                open.push(c);
            } else {
                if(!open.isEmpty() && open.peek() == '(') {
                    min_moves -= 2;
                    open.pop();
                }
            }
        }   

        return min_moves;
    }
}